{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<script>\n",
       "code_show=true; \n",
       "function code_toggle() {\n",
       " if (code_show){\n",
       " $('div.input').hide();\n",
       " } else {\n",
       " $('div.input').show();\n",
       " }\n",
       " code_show = !code_show\n",
       "} \n",
       "$( document ).ready(code_toggle);\n",
       "</script>\n",
       "The raw code for this IPython notebook is by default hidden for easier reading.\n",
       "To toggle on/off the raw code, click <a href=\"javascript:code_toggle()\">here</a>."
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import HTML\n",
    "HTML('''<script>\n",
    "code_show=true; \n",
    "function code_toggle() {\n",
    " if (code_show){\n",
    " $('div.input').hide();\n",
    " } else {\n",
    " $('div.input').show();\n",
    " }\n",
    " code_show = !code_show\n",
    "} \n",
    "$( document ).ready(code_toggle);\n",
    "</script>\n",
    "The raw code for this IPython notebook is by default hidden for easier reading.\n",
    "To toggle on/off the raw code, click <a href=\"javascript:code_toggle()\">here</a>.''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# get the TPM values\n",
    "import pandas as pd\n",
    "tpm_df = pd.read_csv(\"GTEx_Analysis_2016-01-15_v7_RNASeQCv1.1.8_gene_median_tpm.gct\", sep='\\t')\n",
    "gene_name_to_tpm_dict = dict()\n",
    "for index, row in tpm_df.iterrows():\n",
    "    gene_name_to_tpm_dict[row[\"Description\"]] = row[\"Skin - Sun Exposed (Lower leg)\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1230\n"
     ]
    }
   ],
   "source": [
    "# Read in the micro-array data - so far only founds matches between 1230 genes in whitfield in the 56K genes from GTEX\n",
    "# TODO - get better name mapping\n",
    "import math\n",
    "import numpy as np\n",
    "count = 0\n",
    "count_found = 0\n",
    "whitfield_micro_array_df = pd.read_csv(\"Whitfield_2013_cell_cycle_2830_all_scores.csv\")\n",
    "    \n",
    "def get_coarse_phase(angle):\n",
    "    # angles version 2\n",
    "    predicted_phase = -1\n",
    "    if angle >= 0 and angle <= 0.8:\n",
    "        predicted_phase = \"G1\" #G1_angles += 1\n",
    "    elif angle > 0.8 and angle <= 2.5:\n",
    "        predicted_phase = \"S\" #S_angles += 1\n",
    "    elif angle > 2.5 and angle <= 5.4:\n",
    "        predicted_phase = \"G2M\" #G2M_angles += 1\n",
    "    elif angle > 5.4:\n",
    "        predicted_phase = \"G1\" #G1_angles += 1\n",
    "    # angles version 1\n",
    "    #if angle >= 0 and angle <= 2:\n",
    "    #    predicted_phase = \"S\"\n",
    "    #elif angle > 2 and angle <= 4.8:\n",
    "    #    predicted_phase = \"G2M\"\n",
    "    #elif angle > 4.8 and angle <= 5.9:\n",
    "    #    predicted_phase = \"G1\"\n",
    "    #elif angle > 5.9:\n",
    "    #    predicted_phase = \"S\"\n",
    "    return predicted_phase\n",
    "\n",
    "\n",
    "\n",
    "gene_names_with_micro_array_data = list()\n",
    "\n",
    "time_course_1_angles_list = list()\n",
    "time_course_2_angles_list = list()\n",
    "time_course_3_angles_list = list()\n",
    "time_course_4_angles_list = list()\n",
    "time_course_all_phases_list = list()\n",
    "time_course_all_angles_list = list()\n",
    "names_for_all_time_courses_list = list()\n",
    "gene_name_to_time_course_expressions_dict = dict()\n",
    "\n",
    "# calculate the angles for the 4 time courses\n",
    "for i in range(0,40,2): #t-course 1 T = 17.65, φ = 0.0\n",
    "    #print(math.atan2(math.sin((i/17.65)*2*math.pi),math.cos((i/17.65)*2*math.pi))) # with negative radians\n",
    "    angle = ((i/17.65)*2*math.pi)%(2*math.pi)\n",
    "    time_course_1_angles_list.append(angle) # 0 to 2pi\n",
    "    time_course_all_angles_list.append(angle) # 0 to 2pi\n",
    "    time_course_all_phases_list.append(get_coarse_phase(angle)) # 0 to 2pi\n",
    "    names_for_all_time_courses_list.append(\"t1_\" + str(i))\n",
    "for i in range(0,48,2): #t-course 2 T = 18.6, φ = 0.0\n",
    "    #print(math.atan2(math.sin((i/18.6)*2*math.pi),math.cos((i/18.6)*2*math.pi))) # with negative radians\n",
    "    angle = ((i/18.6)*2*math.pi)%(2*math.pi)\n",
    "    time_course_2_angles_list.append(angle) # 0 to 2pi\n",
    "    time_course_all_angles_list.append(angle) # 0 to 2pi\n",
    "    time_course_all_phases_list.append(get_coarse_phase(angle))\n",
    "    names_for_all_time_courses_list.append(\"t2_\" + str(i))\n",
    "for i in range(0,48,2): #t-course 3 T = 18, φ = 0.0\n",
    "    #print(math.atan2(math.sin((i/18)*2*math.pi),math.cos((i/18)*2*math.pi))) # with negative radians\n",
    "    angle = ((i/18)*2*math.pi)%(2*math.pi)\n",
    "    time_course_3_angles_list.append(angle) # 0 to 2pi\n",
    "    time_course_all_angles_list.append(angle) # 0 to 2pi\n",
    "    time_course_all_phases_list.append(get_coarse_phase(angle))\n",
    "    names_for_all_time_courses_list.append(\"t3_\" + str(i))\n",
    "for i in range(0,46,2): #t-course 4 T = 23.95, φ = 2.3\n",
    "    #print(math.atan2(math.sin((i/23.95)*2*math.pi + 2.3),math.cos((i/23.95)*2*math.pi + 2.3))) # with negative radians\n",
    "    angle = ((i/23.95)*2*math.pi + 2.3)%(2*math.pi)\n",
    "    time_course_4_angles_list.append(angle) # 0 to 2pi\n",
    "    time_course_all_angles_list.append(angle) # 0 to 2pi\n",
    "    time_course_all_phases_list.append(get_coarse_phase(angle))\n",
    "    names_for_all_time_courses_list.append(\"t4_\" + str(i))    \n",
    "    \n",
    "    \n",
    "for index, row in whitfield_micro_array_df.iterrows():\n",
    "    count +=1\n",
    "    name = str(row['NAME'])\n",
    "    if (name != 'nan'):\n",
    "        #print(\"___\" + name + \"___\")\n",
    "        names = name.split(\"^\")\n",
    "        if (len(names) > 1):\n",
    "            #print(\"___\" + names[1] + \"___\")\n",
    "            gene_name = names[1].strip()\n",
    "            if (gene_name in gene_name_to_tpm_dict and not(gene_name in gene_names_with_micro_array_data)):\n",
    "                \n",
    "                \n",
    "                ####!!!! cut-off to have just a few top cycling genes\n",
    "                scaled_fourier_score = float(row['Scaled Fourier'])\n",
    "                #if (scaled_fourier_score < 10.0):\n",
    "                 #   continue\n",
    "                    \n",
    "                tpm = gene_name_to_tpm_dict[gene_name]\n",
    "                #print(gene_name + \" is found\")\n",
    "                gene_names_with_micro_array_data.append(gene_name)\n",
    "                count_found += 1\n",
    "                gene_name_to_time_course_expressions_dict[gene_name] = dict()\n",
    "                gene_name_to_time_course_expressions_dict[gene_name][\"all\"] = list()\n",
    "                gene_name_to_time_course_expressions_dict[gene_name][\"one\"] = list()\n",
    "                \n",
    "                \n",
    "                new_row = row[11:31].replace(r'\\s+', np.nan, regex=True)\n",
    "                new_row = new_row.fillna(0)\n",
    "                row[11:31] = new_row\n",
    "            \n",
    "                new_row = row[32:56].replace(r'\\s+', np.nan, regex=True)\n",
    "                new_row = new_row.fillna(0)\n",
    "                row[32:56] = new_row\n",
    "                \n",
    "                new_row = row[57:81].replace(r'\\s+', np.nan, regex=True)\n",
    "                new_row = new_row.fillna(0)\n",
    "                row[57:81] = new_row\n",
    "                \n",
    "                new_row = row[82:105].replace(r'\\s+', np.nan, regex=True)\n",
    "                new_row = new_row.fillna(0)\n",
    "                row[82:105] = new_row\n",
    "                \n",
    "                \n",
    "                #TODO: for now adding zero to missing time courses and samples - valid for all things?\n",
    "                \n",
    "                for i in range(11,31):\n",
    "                    gene_name_to_time_course_expressions_dict[gene_name][\"one\"].append((2**row[i]) * tpm if row[i] else 0) # undo the log2 of data\n",
    "                    gene_name_to_time_course_expressions_dict[gene_name][\"all\"].append((2**row[i]) * tpm if row[i] else 0) # undo the log2 of data\n",
    "                gene_name_to_time_course_expressions_dict[gene_name][\"two\"] = list()\n",
    "                for i in range(32,56):\n",
    "                    gene_name_to_time_course_expressions_dict[gene_name][\"all\"].append((2**row[i]) * tpm if row[i] else 0)\n",
    "                    gene_name_to_time_course_expressions_dict[gene_name][\"two\"].append((2**row[i]) * tpm if row[i] else 0) # undo the log2 of data\n",
    "                gene_name_to_time_course_expressions_dict[gene_name][\"three\"] = list()\n",
    "                for i in range(57,81):\n",
    "                    gene_name_to_time_course_expressions_dict[gene_name][\"all\"].append((2**row[i]) * tpm if row[i] else 0)\n",
    "                    gene_name_to_time_course_expressions_dict[gene_name][\"three\"].append((2**row[i]) * tpm if row[i] else 0) # undo the log2 of data\n",
    "                gene_name_to_time_course_expressions_dict[gene_name][\"four\"] = list()\n",
    "                for i in range(82,105):\n",
    "                    gene_name_to_time_course_expressions_dict[gene_name][\"all\"].append((2**row[i]) * tpm if row[i] else 0)\n",
    "                    gene_name_to_time_course_expressions_dict[gene_name][\"four\"].append((2**row[i]) * tpm if row[i] else 0) # undo the log2 of data\n",
    "    \n",
    "                gene_name_to_time_course_expressions_dict[gene_name][\"sine\"] = row['Sin']\n",
    "                gene_name_to_time_course_expressions_dict[gene_name][\"cosine\"] = row['Cos']\n",
    "                gene_name_to_time_course_expressions_dict[gene_name][\"phase\"] = row['Phase'].strip()\n",
    "                gene_name_to_time_course_expressions_dict[gene_name][\"fourier_score\"] = scaled_fourier_score\n",
    "print(count_found)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['class_labels', 'counts', 'ensIds', 'labels', 'sym_names']\n",
      "[1. 2. 4. 3.]\n",
      "38293\n",
      "(3, 38293)\n",
      "(38050, 3)\n"
     ]
    }
   ],
   "source": [
    "# data_mESCbulk.h5f 1:1 RUN this cell\n",
    "import h5py\n",
    "import pandas as pd\n",
    "\n",
    "SC_df = pd.DataFrame()\n",
    "ftst = h5py.File(\"data_mESCbulk.h5f\",'r')\n",
    "print([key for key in ftst.keys()])\n",
    "print(ftst['labels'][:])\n",
    "gene_names_all = ftst['sym_names'][:]\n",
    "gene_names_all = [x.decode('utf-8').upper() for x in gene_names_all]\n",
    "print(len(gene_names_all))\n",
    "YT = ftst['counts'][:]\n",
    "YT = YT[[0,1,3],:]\n",
    "print(YT.shape)\n",
    "labels = [\"bulk_G1\", \"bulk_S\", \"bulk_G2M\"]\n",
    "\n",
    "SC_df = pd.DataFrame(YT.T)\n",
    "SC_df.columns = labels\n",
    "SC_df.index = gene_names_all\n",
    "SC_df = SC_df[~SC_df.index.duplicated(keep='first')]\n",
    "print(SC_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['KG1', 'KG2M', 'KS', 'K_phase', 'LogNcountsQuartz', 'LogVar_techQuartz_logfit', 'ccCBall_gene_indices', 'ccGO_gene_indices', 'cell_names', 'cellcyclegenesGO', 'gene_names_all', 'genes_heterogen', 'phase_vecG1', 'phase_vecG2M', 'phase_vecS', 'sym_namesQ']\n",
      "[b'ES_G1_1' b'ES_G1_2' b'ES_G1_3' b'ES_G1_4' b'ES_G1_5' b'ES_G1_6'\n",
      " b'ES_G1_7' b'ES_G1_8' b'ES_G1_9' b'ES_G1_10' b'ES_G1_11' b'ES_G1_12'\n",
      " b'ES_G1_13' b'ES_G1_14' b'ES_G1_15' b'ES_G1_16' b'ES_G1_17' b'ES_G1_18'\n",
      " b'ES_G1_19' b'ES_G1_20' b'ES_S_1' b'ES_S_2' b'ES_S_3' b'ES_S_4' b'ES_S_5'\n",
      " b'ES_S_6' b'ES_S_7' b'ES_G2M_1' b'ES_G2M_2' b'ES_G2M_3' b'ES_G2M_4'\n",
      " b'ES_G2M_5' b'ES_G2M_6' b'ES_G2M_7' b'ES_G2M_8']\n",
      "[1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 3 3 3 3 3 3 3 3]\n"
     ]
    }
   ],
   "source": [
    "# MouseEST 35 1:3 RUN this cell for mouseEST\n",
    "import numpy as np\n",
    "import h5py\n",
    "ftst = h5py.File(\"normCounts_mESCquartz.h5f\",'r')\n",
    "print([key for key in ftst.keys()])\n",
    "print(ftst['cell_names'][:])\n",
    "labels_tst = ftst['phase_vecS'][:]*2+ftst['phase_vecG2M'][:]*3+ftst['phase_vecG1'][:]\n",
    "labels_tst = np.array([int(i) for i in labels_tst])\n",
    "print(labels_tst)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "querying 1-1000...done.\n",
      "querying 1001-2000...done.\n",
      "querying 2001-3000...done.\n",
      "querying 3001-4000...done.\n",
      "querying 4001-5000...done.\n",
      "querying 5001-6000...done.\n",
      "querying 6001-7000...done.\n",
      "querying 7001-8000...done.\n",
      "querying 8001-9000...done.\n",
      "querying 9001-10000...done.\n",
      "querying 10001-11000...done.\n",
      "querying 11001-12000...done.\n",
      "querying 12001-13000...done.\n",
      "querying 13001-14000...done.\n",
      "querying 14001-15000...done.\n",
      "querying 15001-16000...done.\n",
      "querying 16001-17000...done.\n",
      "querying 17001-18000...done.\n",
      "querying 18001-19000...done.\n",
      "querying 19001-20000...done.\n",
      "querying 20001-21000...done.\n",
      "querying 21001-22000...done.\n",
      "querying 22001-23000...done.\n",
      "querying 23001-24000...done.\n",
      "querying 24001-25000...done.\n",
      "querying 25001-26000...done.\n",
      "querying 26001-27000...done.\n",
      "querying 27001-28000...done.\n",
      "querying 28001-29000...done.\n",
      "querying 29001-30000...done.\n",
      "querying 30001-31000...done.\n",
      "querying 31001-32000...done.\n",
      "querying 32001-33000...done.\n",
      "querying 33001-34000...done.\n",
      "querying 34001-35000...done.\n",
      "querying 35001-36000...done.\n",
      "querying 36001-36807...done.\n",
      "Finished.\n",
      "4932 input query terms found no hit:\n",
      "\t['ENSMUSG00000000702', 'ENSMUSG00000001379', 'ENSMUSG00000001417', 'ENSMUSG00000002290', 'ENSMUSG000\n",
      "Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n",
      "(35, 36807)\n",
      "[1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 2. 2. 2. 2.\n",
      " 2. 2. 2. 3. 3. 3. 3. 3. 3. 3. 3.]\n",
      "[b'ES_G1_1' b'ES_G1_2' b'ES_G1_3' b'ES_G1_4' b'ES_G1_5' b'ES_G1_6'\n",
      " b'ES_G1_7' b'ES_G1_8' b'ES_G1_9' b'ES_G1_10' b'ES_G1_11' b'ES_G1_12'\n",
      " b'ES_G1_13' b'ES_G1_14' b'ES_G1_15' b'ES_G1_16' b'ES_G1_17' b'ES_G1_18'\n",
      " b'ES_G1_19' b'ES_G1_20' b'ES_S_1' b'ES_S_2' b'ES_S_3' b'ES_S_4' b'ES_S_5'\n",
      " b'ES_S_6' b'ES_S_7' b'ES_G2M_1' b'ES_G2M_2' b'ES_G2M_3' b'ES_G2M_4'\n",
      " b'ES_G2M_5' b'ES_G2M_6' b'ES_G2M_7' b'ES_G2M_8']\n"
     ]
    }
   ],
   "source": [
    "# MouseEST 35 2:3 RUN this cell for mouseEST\n",
    "# a bit slow, try not to re-run\n",
    "import mygene\n",
    "mg = mygene.MyGeneInfo()\n",
    "\n",
    "gene_names_all = ftst['gene_names_all'][:]\n",
    "gene_names_all = [x.decode('utf-8') for x in gene_names_all]\n",
    "geneSyms = mg.querymany(gene_names_all, scopes='ensembl.gene', fields='symbol', species='mouse')\n",
    "newGeneList = list()\n",
    "for i in range(0, len(gene_names_all)):\n",
    "    if 'symbol' in geneSyms[i]:\n",
    "        newGeneList.append(geneSyms[i]['symbol'].upper())\n",
    "    else:\n",
    "        newGeneList.append(gene_names_all[i])\n",
    "\n",
    "YT = ftst['LogNcountsQuartz'][:]\n",
    "print(YT.shape)\n",
    "labels_tst = ftst['phase_vecS'][:]*2+ftst['phase_vecG2M'][:]*3+ftst['phase_vecG1'][:]\n",
    "print(labels_tst)\n",
    "print(ftst[\"cell_names\"][:])\n",
    "# G1, S, G2M"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(36807, 35)\n",
      "(36793, 35)\n",
      "0.9468431505995761\n"
     ]
    }
   ],
   "source": [
    "# MouseEST 35 3:3 RUN this cell for mouseEST\n",
    "string_labels = list()\n",
    "cell_names = ftst[\"cell_names\"][:]\n",
    "cell_names = gene_names_all = [x.decode('utf-8') for x in cell_names]\n",
    "for label in cell_names:\n",
    "    if (\"G1\" in label):\n",
    "        string_labels.append(\"G1\")\n",
    "    if (\"S\" in label):\n",
    "        string_labels.append(\"S\")\n",
    "    if (\"G2\" in label):\n",
    "        string_labels.append(\"G2\")\n",
    "SC_df = pd.DataFrame(YT.T)\n",
    "SC_df.columns = cell_names\n",
    "SC_df.index = newGeneList\n",
    "\n",
    "print(SC_df.shape)\n",
    "SC_df = SC_df[~SC_df.index.duplicated(keep='first')]\n",
    "print(SC_df.shape)\n",
    "print(SC_df[\"ES_G1_1\"][\"BFAR\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "90\n",
      "81\n"
     ]
    }
   ],
   "source": [
    "# Stahlberg 1:1 RUN this cell for Stahlberg\n",
    "import pandas as pd\n",
    "\n",
    "# read in the single-cell data, and identify the set of genes that also have micro-array/tpm expression (1114)\n",
    "# for each cell, get percentage of reads for each gene\n",
    "# for the micro-array set, get percentage of \"expression\" for each gene\n",
    "# for each gene, get the best match of percentages for each of the time-courses, and take an average angle\n",
    "# then take the average angle of all the genes - that gives the phase\n",
    "### process SC single cell data for Seurat input\n",
    "SC_df = pd.read_csv(\"Stahlberg_MCF7_group_1_to_6.csv\", sep=',')\n",
    "def phase_to_string(phase_number):\n",
    "    if (phase_number in (1, 4, 9)):\n",
    "        return \"G1\"\n",
    "    if (phase_number in (2, 5, 10)):\n",
    "        return \"S\"\n",
    "    if (phase_number in (3, 6, 11)):\n",
    "        return \"G2M\"\n",
    "SC_df = SC_df.rename(lambda x: x + \"_\" + phase_to_string(int(SC_df.loc[x, \"#Group\"])))\n",
    "SC_df = SC_df.iloc[:,:-3]\n",
    "SC_df = SC_df.replace(-1,0.0).T\n",
    "print(len(SC_df.columns))\n",
    "print(len(SC_df.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['KIF20A', 'HSPA1A', 'PLK1', 'MYB', 'CCNB1', 'CCNE2', 'CDCA8', 'SGOL2',\n",
      "       'ASPM', 'TROAP',\n",
      "       ...\n",
      "       'TM4SF1', 'PER1', 'FAM3C', 'CKS1B', 'GUCY1B3', 'APOBEC3C', 'AMICA1',\n",
      "       'BCL11B', 'PPP2R5E', 'DDX11'],\n",
      "      dtype='object', length=1230)\n",
      "Index(['t1_0_G1', 't1_2_G1', 't1_4_S', 't1_6_S', 't1_8_G2M', 't1_10_G2M',\n",
      "       't1_12_G2M', 't1_14_G2M', 't1_16_G1', 't1_18_G1', 't1_20_S', 't1_22_S',\n",
      "       't1_24_S', 't1_26_G2M', 't1_28_G2M', 't1_30_G2M', 't1_32_G2M',\n",
      "       't1_34_G1', 't1_36_G1', 't1_38_S', 't2_0_G1', 't2_2_G1', 't2_4_S',\n",
      "       't2_6_S', 't2_8_G2M', 't2_10_G2M', 't2_12_G2M', 't2_14_G2M', 't2_16_G1',\n",
      "       't2_18_G1', 't2_20_G1', 't2_22_S', 't2_24_S', 't2_26_S', 't2_28_G2M',\n",
      "       't2_30_G2M', 't2_32_G2M', 't2_34_G2M', 't2_36_G1', 't2_38_G1',\n",
      "       't2_40_S', 't2_42_S', 't2_44_S', 't2_46_G2M', 't3_0_G1', 't3_2_G1',\n",
      "       't3_4_S', 't3_6_S', 't3_8_G2M', 't3_10_G2M', 't3_12_G2M', 't3_14_G2M',\n",
      "       't3_16_G1', 't3_18_G1', 't3_20_G1', 't3_22_S', 't3_24_S', 't3_26_G2M',\n",
      "       't3_28_G2M', 't3_30_G2M', 't3_32_G2M', 't3_34_G1', 't3_36_G1',\n",
      "       't3_38_G1', 't3_40_S', 't3_42_S', 't3_44_G2M', 't3_46_G2M', 't4_0_S',\n",
      "       't4_2_G2M', 't4_4_G2M', 't4_6_G2M', 't4_8_G2M', 't4_10_G2M', 't4_12_G1',\n",
      "       't4_14_G1', 't4_16_G1', 't4_18_G1', 't4_20_S', 't4_22_S', 't4_24_S',\n",
      "       't4_26_G2M', 't4_28_G2M', 't4_30_G2M', 't4_32_G2M', 't4_34_G2M',\n",
      "       't4_36_G1', 't4_38_G1', 't4_40_G1', 't4_42_G1', 't4_44_S'],\n",
      "      dtype='object')\n",
      "['KIF20A', 'HSPA1A', 'PLK1', 'MYB', 'CCNB1', 'CCNE2', 'CDCA8', 'SGOL2', 'ASPM', 'TROAP', 'CENPE', 'BUB1', 'HIST1H4C', 'E2F2', 'CDCA2', 'HIST1H4L', 'HIST1H1A', 'NEK2', 'ARHGAP11A', 'TOP2A', 'HIST1H4B', 'KPNA2', 'CDO1', 'STARD8', 'TTK', 'HIST1H4D', 'KIF23', 'GPSM2', 'CDC25C', 'NDE1', 'CDC20', 'CENPA', 'DLGAP1', 'CENPF', 'CKS2', 'KIF2C', 'UBE2C', 'GTSE1', 'AREG', 'CKAP2', 'CCNB2', 'SPAG5', 'DEPDC1', 'KIF18A', 'LZTS1', 'BCL11A', 'TRIM59', 'ZNF587', 'MKI67', 'DEPDC1B', 'TPX2', 'CDCA3', 'HSPA8', 'HMGA2', 'RAB39B', 'ECT2', 'CD200', 'BBC3', 'UNG', 'CDC6', 'CCNE1', 'HMMR', 'JMY', 'PRKAA1', 'CCNA2', 'PRC1', 'RAD21', 'MCM2', 'PLCXD1', 'CCNF', 'BUB1B', 'FZD8', 'SSBP2', 'HIST1H4E', 'MAF', 'CHAF1B', 'SP8', 'PCNA', 'FST', 'PSMD11', 'VMP1', 'VANGL1', 'TOR3A', 'MSH6', 'LIPE', 'USP47', 'SYTL4', 'HN1', 'DYRK2', 'MLYCD', 'RACGAP1', 'FBXO32', 'YPEL4', 'NEDD4L', 'ARRDC3', 'RASEF', 'ATF3', 'ATAD2', 'HMGB3', 'CDC25B', 'BAG3', 'HSPH1', 'TREX1', 'CLSPN', 'KIF4A', 'TRIM45', 'CISH', 'CDKN1A', 'RFXAP', 'ZNF395', 'CXXC4', 'MNT', 'ZNF165', 'RAB11FIP1', 'RNF26', 'ATF7IP2', 'SESN2', 'PLK3', 'KIF11', 'TNFAIP8', 'RABGEF1', 'MYH11', 'FAM46A', 'TRIB1', 'NR1D2', 'ESPL1', 'CREB5', 'PPM1F', 'SMTN', 'RBM12', 'ZMYM1', 'OXGR1', 'BCL2L11', 'NINJ1', 'CD97', 'NR4A2', 'MESDC1', 'TP53TG3', 'CDKN1C', 'MCM6', 'RORA', 'RBM17', 'MCM4', 'PTGS2', 'ANLN', 'DNAJA1', 'SLC25A29', 'GPRC5C', 'ARMC8', 'SP1', 'PPP2R2A', 'YEATS4', 'TTYH2', 'SLC1A5', 'CDKN3', 'NUDT4', 'LRIG3', 'MCM10', 'RBBP8', 'JUNB', 'RUNX1', 'SLC16A7', 'EPHB3', 'CDR2', 'IL8', 'FGF5', 'FAM20C', 'FANCL', 'UBE2S', 'TULP3', 'CDT1', 'CTDSP1', 'POLD3', 'NUSAP1', 'HOXA7', 'CBLB', 'SLCO1B3', 'TES', 'SUV420H1', 'LAMA5', 'HIST1H2AJ', 'ACACA', 'TLR3', 'SF1', 'CAMK4', 'STK17B', 'MAT2A', 'GRPEL2', 'LAG3', 'C8orf4', 'KCNA1', 'BPGM', 'NR0B1', 'NAV2', 'SYT15', 'MAFK', 'MTMR6', 'PLSCR4', 'DNAJB1', 'GLCCI1', 'USP37', 'PELI2', 'DNAJA4', 'GAS2L3', 'HIST2H2AB', 'RFC4', 'ADAT1', 'TSLP', 'BIRC3', 'FANCE', 'HIST2H2AC', 'SLBP', 'RPL28', 'ZDHHC23', 'ARHGAP19', 'OGT', 'ZNF135', 'HSPA1L', 'DDX52', 'CYP26B1', 'ARID3B', 'TACC3', 'IRX4', 'PRKD1', 'BTRC', 'PCDH9', 'RPS6KA5', 'PDGFRA', 'FGF21', 'SPAG16', 'ARHGAP25', 'SLC4A11', 'PLD1', 'FGF9', 'HOXA4', 'PTTG1', 'RAB9B', 'PKNOX1', 'CDC42EP2', 'CHAF1A', 'CFLAR', 'SHCBP1', 'GOT1', 'PREX1', 'KCTD9', 'UBE2D3', 'ANKRD28', 'SMPDL3A', 'CHRNA10', 'GABPB2', 'ASCC3', 'FOXD4', 'UBQLN1', 'P2RY2', 'DCP2', 'MOSPD2', 'CDKN2D', 'RASSF1', 'OSBPL6', 'BTN2A2', 'RRM2', 'PEG3', 'HIST1H2BD', 'IL24', 'EXO1', 'HSPD1', 'KRTAP4-7', 'SMARCA2', 'DTNA', 'PTTG2', 'TRAF6', 'CHDH', 'ANKRA2', 'NUP98', 'MNS1', 'PDGFA', 'PTPN13', 'GIT2', 'RAD51', 'C1GALT1', 'PLEKHF1', 'MCM3', 'KIF26A', 'CCNK', 'HMGCR', 'SLC39A8', 'IRF2BP2', 'PLAU', 'C21orf119', 'PAFAH1B1', 'ARHGAP20', 'TP53INP1', 'ATP1B1', 'DDIT3', 'SNX22', 'CELSR1', 'TTPA', 'PCDH8', 'GUCY1A2', 'NAV3', 'PVRL4', 'PRSS35', 'NOTCH2', 'EVI2A', 'ZIC3', 'TWISTNB', 'CTSK', 'JUND', 'DDX17', 'HIST3H2A', 'SATB1', 'DAAM1', 'CALM2', 'DDAH1', 'MYO10', 'ZBED3', 'CSTF1', 'KLF2', 'GADD45A', 'SLC25A27', 'CXADR', 'CDC25A', 'SURF6', 'MYLK2', 'CCND3', 'KLF5', 'HMGB2', 'FRAT2', 'NR3C2', 'USP53', 'PAQR4', 'MRC2', 'MEIS2', 'GLRB', 'PHKA1', 'PSIP1', 'HECA', 'NUP37', 'SNAPC1', 'KLF9', 'STK38L', 'RNF152', 'ING5', 'ARMC7', 'PMAIP1', 'TGFB3', 'TCF19', 'OLIG1', 'MARK1', 'LACTB2', 'WASL', 'LGALS8', 'GOLPH3', 'TRIM2', 'TBC1D8', 'PARP8', 'KIRREL2', 'SACS', 'KHDRBS1', 'OAZ3', 'PGPEP1', 'GPR126', 'NCOA7', 'NDUFC2', 'NCOA2', 'ANTXR2', 'HOXB13', 'CIT', 'ARMC4', 'ZNF480', 'WHSC1L1', 'ADAMTS3', 'ZRANB3', 'IRX3', 'RTN4R', 'RAB4B', 'GATA4', 'INCENP', 'YPEL2', 'WDR3', 'ABTB1', 'PRKAA2', 'ZNF71', 'TOP1', 'LIMK2', 'CEBPA', 'TGFA', 'SFRP4', 'CHD7', 'NAV1', 'KCNJ14', 'GPD2', 'PXK', 'NBR2', 'TRAF1', 'MAP3K5', 'IL23A', 'ARIH2', 'USP51', 'GRASP', 'KIF22', 'ADCK2', 'HIST1H1E', 'SLC38A2', 'SLC2A4', 'NFIA', 'MBNL2', 'ZNF202', 'KIF5B', 'CHEK1', 'GSC', 'CASP6', 'PLEKHA5', 'CHST6', 'LCN1', 'HELLS', 'HSPA5', 'HIST1H3B', 'BRIP1', 'RHEB', 'INSR', 'CDH24', 'ARL4A', 'TBL2', 'BBX', 'ATP2A2', 'DISC1', 'SYNCRIP', 'ENPP5', 'PLK4', 'PAIP1', 'CABYR', 'NRF1', 'UNC119', 'FBXO5', 'EIF4E', 'ETS2', 'KCTD8', 'ZNRF2', 'ARL6IP6', 'BRCA1', 'BRI3BP', 'MOAP1', 'MTHFD2', 'MAN2C1', 'ANKRD26', 'HOXD8', 'SLCO4A1', 'ZNF33A', 'IL7', 'CXorf23', 'KIAA0101', 'CASP8AP2', 'TCFL5', 'STAM2', 'IRF1', 'CYHR1', 'REL', 'LBR', 'AP2A2', 'PASK', 'MLLT10', 'CASP2', 'WDHD1', 'NOXA1', 'NAP1L5', 'TPPP', 'THUMPD2', 'ULK1', 'ANXA3', 'RPS6KB1', 'AHNAK', 'WDR26', 'IGSF3', 'DMRTA1', 'MAFG', 'FH', 'RIN2', 'PURA', 'PRIM1', 'MASTL', 'KIAA1199', 'TAF4', 'CYP1A1', 'RAB8B', 'RAD54L', 'ZNF26', 'ENPP4', 'TDP1', 'PLA1A', 'ABCG4', 'SQLE', 'BCAR3', 'OIP5', 'TIGD5', 'DONSON', 'CPNE8', 'KIAA0247', 'ZNF467', 'KIN', 'EAF2', 'LIN9', 'EVI5', 'DLX4', 'DNM3', 'DCLRE1B', 'CBFB', 'HIST1H2AM', 'KIAA0408', 'MAPKBP1', 'NFIB', 'UGT8', 'ANP32E', 'CPM', 'DUSP1', 'BIN1', 'USP43', 'TXNRD1', 'ZNF644', 'AMIGO2', 'ADAMTS1', 'ZFPM2', 'RCOR1', 'EDA', 'PHLDA1', 'HOXA5', 'PSG2', 'ICMT', 'WDR1', 'BBS4', 'SHOX2', 'LMO7', 'ZNF548', 'PPIE', 'GRM8', 'AP1S1', 'DNAJB6', 'BMF', 'BANP', 'SYNE2', 'GSTT2', 'PPP1R16B', 'RHOBTB1', 'CTNND1', 'CYR61', 'ZNF620', 'TNFRSF25', 'LIPH', 'AP1G1', 'C1QTNF2', 'ZNF324', 'EIF5', 'GNPNAT1', 'SDC1', 'ADCY6', 'ITK', 'SPON2', 'RRAD', 'MGAT5', 'RAPGEF6', 'UHRF1', 'THAP7', 'PIK3C2B', 'F11R', 'DOCK7', 'HLCS', 'SLC7A2', 'HSPA6', 'SLC25A5', 'SESN3', 'TBC1D5', 'HOXA10', 'SGOL1', 'HOXD9', 'CTNNBIP1', 'PCSK4', 'CTTNBP2', 'PRSS16', 'BRD8', 'PCBP2', 'TRIM36', 'NPFF', 'TLR6', 'BTG3', 'ZNF589', 'SLC25A24', 'ACN9', 'PPARGC1B', 'DRAP1', 'PPFIA4', 'LAMC1', 'SCN5A', 'SLC25A21', 'GDAP1', 'EYA3', 'SFPQ', 'RNF4', 'GREB1', 'SLC4A7', 'HOXB8', 'GLI3', 'APPBP2', 'HIST1H2AD', 'MESP1', 'FZD1', 'LRP10', 'CLCN2', 'ZNF514', 'PPP2R5C', 'PCDHB12', 'HOXC13', 'HUS1B', 'KCNE1L', 'ZNF614', 'PLAG1', 'SLC1A3', 'BCL6', 'IL17D', 'SMPD3', 'GPAM', 'CAPS2', 'BMP7', 'MAML1', 'PARD6G', 'NET1', 'SNCG', 'GLS', 'CYP2R1', 'PTP4A1', 'TSC1', 'SNX10', 'ALPK1', 'BRD3', 'SRD5A2', 'MAPK11', 'PARP14', 'RFC1', 'MCM5', 'TOP3A', 'CYP2U1', 'PPM1D', 'TP53RK', 'MEF2C', 'SLC2A4RG', 'BTN3A3', 'GTPBP2', 'B2M', 'OTX1', 'NPM1', 'SERPINH1', 'PPP1CB', 'CALU', 'RAB27A', 'KNTC1', 'OXR1', 'NRTN', 'ABHD4', 'UBE3A', 'MAGEH1', 'NKX2-2', 'UBE2N', 'AADAC', 'POLG2', 'DSP', 'SLC9A3R1', 'CCR7', 'GGA1', 'SPATA5L1', 'COL27A1', 'FNBP4', 'LMO4', 'KCNC1', 'APBB3', 'SMAD6', 'SPAG1', 'APOL6', 'TIMM17B', 'AKAP5', 'TRIM52', 'ANXA1', 'RAC2', 'TULP4', 'SPIN3', 'MUM1L1', 'FBXO22', 'GAB1', 'PHLDB2', 'NCOA3', 'RNF103', 'HSF2', 'IGSF9', 'RFX3', 'ADORA2B', 'DTX3L', 'AMPH', 'GK', 'PGBD5', 'WSB1', 'YTHDF2', 'NANOG', 'ACSL4', 'CECR6', 'BTBD3', 'JPH3', 'HCN3', 'SPIRE2', 'EPB41L4B', 'LDLR', 'EPB41L5', 'COL4A4', 'FIGNL1', 'CASC2', 'TUBD1', 'SRRM2', 'DOCK11', 'NUP50', 'PADI1', 'CHST7', 'PML', 'CARD14', 'OSBPL10', 'KLF13', 'TBL1Y', 'MXD4', 'GALNT13', 'SRCAP', 'PTPN18', 'ALG10', 'THAP8', 'RHOB', 'EHD1', 'FZR1', 'KIAA0586', 'PTHLH', 'IRX5', 'UBE2G1', 'LYN', 'PRPS2', 'ZFYVE1', 'PBXIP1', 'TRIM40', 'CLK1', 'BIRC5', 'CHIC2', 'GTF2E1', 'UBE2D1', 'LCMT2', 'PLEKHG3', 'PNMA3', 'PEX6', 'SNAP23', 'ADCY1', 'DHX32', 'PARP9', 'AVIL', 'PEX13', 'ZHX2', 'BUB3', 'VIPR1', 'CDYL', 'PSCA', 'DNAJB4', 'GHR', 'OR4N4', 'WFS1', 'MERTK', 'NUP35', 'ZNF207', 'TAF15', 'HLA-DQB2', 'TLK2', 'HOXB9', 'RSBN1', 'PRICKLE2', 'KLF7', 'ORM2', 'ARHGEF10', 'VRK1', 'CTRL', 'H2AFX', 'EPC1', 'TAF1B', 'PLAGL1', 'CTNNA1', 'ORAOV1', 'ZMYND19', 'ARPC5L', 'WDR36', 'OCA2', 'AZI2', 'ZBTB11', 'PKMYT1', 'BAALC', 'SIGLEC6', 'C6orf62', 'CSTF2T', 'SESTD1', 'FGFRL1', 'SOAT2', 'E2F7', 'MAN1A2', 'PRKCE', 'RAD1', 'C11orf30', 'PIM3', 'MNDA', 'ICA1', 'KLHL18', 'BAI1', 'CNOT4', 'MANEA', 'AP2B1', 'ID2', 'ZADH2', 'TNFAIP1', 'PDE5A', 'MIB1', 'AMD1', 'HIST1H2AH', 'GOLPH3L', 'PSD3', 'FGF12', 'MKNK2', 'SLC4A4', 'SLC2A12', 'SH3BP2', 'TRAF3', 'TNFAIP2', 'SLC35E3', 'UBTF', 'PSAP', 'CLK4', 'DCLRE1C', 'MDM1', 'RBMS3', 'CDH18', 'TOB2', 'RAB1A', 'JAK1', 'RIMS4', 'EFNA3', 'NAPG', 'TMPO', 'WWTR1', 'CNNM3', 'FAM35A', 'IVNS1ABP', 'RALBP1', 'PVRL2', 'TRIM56', 'AP3M2', 'MAPRE2', 'KPNA4', 'HIPK3', 'CROT', 'RAB31', 'MAN1A1', 'SPP2', 'FOSL2', 'WEE1', 'GTF2E2', 'MTMR4', 'ITGB8', 'PYGO1', 'MYO1E', 'PRRX2', 'ARHGEF19', 'PRAF2', 'CSAD', 'ACSL3', 'KIFAP3', 'ASH1L', 'PTEN', 'RABIF', 'B3GALT6', 'FN3KRP', 'SLCO2A1', 'GPRASP2', 'C9orf116', 'ID4', 'METTL4', 'COX7B', 'RNF41', 'OPA1', 'KLF11', 'THBS2', 'DNAJC3', 'KIAA1524', 'SF3B1', 'ITGAX', 'DCBLD2', 'TIFA', 'TOMM20', 'CTSZ', 'UBAP2L', 'PPP2R5D', 'KIAA1468', 'STXBP2', 'PNRC2', 'DGAT2', 'DUSP10', 'TACC1', 'PPP1R9B', 'CTNNB1', 'E2F1', 'SNRPN', 'NUMB', 'F2RL2', 'MATN1', 'LHX6', 'DDX10', 'RCBTB1', 'TJP1', 'KIF3C', 'HHAT', 'ASPH', 'HSPA4', 'APC2', 'CDH10', 'MRPS23', 'PPAP2C', 'RGS3', 'ZNF610', 'PRDM10', 'WWOX', 'YWHAZ', 'SPATA9', 'ELAVL2', 'PPP1R3E', 'HDAC8', 'SEL1L', 'VPS29', 'SMARCC2', 'COQ7', 'ZNF264', 'FGFR2', 'SMPD1', 'WFDC12', 'SYNE1', 'CEACAM1', 'KLF6', 'MRAS', 'PHF2', 'BRCA2', 'BCL7A', 'MGEA5', 'DEFB103A', 'MTR', 'CDKN2C', 'TCIRG1', 'THOC2', 'PTTG1IP', 'MPP1', 'KIF13B', 'TTC16', 'STAR', 'PSMD7', 'NFYB', 'GNE', 'ZNF519', 'USP45', 'MTCP1', 'MAGEE1', 'HPS4', 'MYH3', 'USP18', 'KCNA6', 'SNAP29', 'FNDC5', 'FGF13', 'MSX1', 'CHKA', 'C19orf25', 'EXOC8', 'CDC27', 'E2F5', 'GNAI1', 'ZNF580', 'SHMT1', 'MAGEL2', 'SAMD9', 'PAPD4', 'CAPN14', 'RGS16', 'PBX1', 'SLC1A1', 'ZNF429', 'RUVBL1', 'CA2', 'CPA3', 'UST', 'RGS17', 'PLA2G12A', 'IER5', 'PNPT1', 'SERP1', 'CALD1', 'TRIM68', 'SLC25A16', 'HOXA3', 'PALMD', 'AMY2B', 'NOTCH4', 'MTM1', 'TIMM23', 'ZNF22', 'SP3', 'PTPRE', 'CLDN12', 'IGF2R', 'GABRQ', 'RPIA', 'CARD9', 'MPV17', 'PKIA', 'RAB35', 'ASB7', 'ZNF484', 'RBM14', 'RNF8', 'BAI3', 'RRS1', 'DPCR1', 'ADAMTSL3', 'FAM8A1', 'DGCR8', 'LIFR', 'MYO6', 'BARD1', 'RBM6', 'PTPN12', 'WTAP', 'NDRG4', 'ZNF24', 'MAP1B', 'CACNA1C', 'DZIP3', 'SEC14L1', 'WBP5', 'ENC1', 'KRTAP19-1', 'BRD4', 'TAF1A', 'TSPYL4', 'C9orf37', 'KCNC3', 'MECP2', 'CHD9', 'SON', 'SERTAD3', 'PANK2', 'CSNK1D', 'PROX1', 'MTRF1L', 'ZC3HAV1', 'MSH2', 'NLGN4X', 'ZDHHC14', 'UAP1', 'PPP3CC', 'U2AF2', 'SENP6', 'ASB13', 'KIAA1211', 'GOLGA4', 'DLG1', 'DHFR', 'TAF7', 'CDC40', 'ZNF561', 'SIX4', 'MED12L', 'TCEAL7', 'CRIP2', 'PGM3', 'RAD23B', 'MAGED2', 'MFAP3', 'BCL2', 'CCIN', 'FDFT1', 'RBM15', 'PITPNB', 'NSFL1C', 'BTBD7', 'ATF5', 'RFC2', 'KIF21A', 'GMIP', 'BDKRB1', 'ADRB2', 'EMILIN2', 'GUCA1B', 'ANKRD9', 'PHKG1', 'PVALB', 'PHF6', 'HEY1', 'RORC', 'SPAG9', 'FANK1', 'HECTD1', 'ZHX3', 'ZNF134', 'APOBEC3F', 'EFNA1', 'RPA3', 'GRIA1', 'KCNG3', 'TCF3', 'DLGAP4', 'MLF1IP', 'PAEP', 'RABL3', 'RAB3IL1', 'DPF3', 'CDKL2', 'BAIAP2', 'DUSP6', 'ACYP1', 'FAIM2', 'AMOTL1', 'FCN1', 'ANGPT2', 'IL4', 'ZNF572', 'OTUD1', 'AHI1', 'FBXO4', 'BCORL1', 'NOX5', 'COL4A1', 'RNPS1', 'LMO2', 'LRRC20', 'KLKB1', 'KLHL7', 'HSPB8', 'SYNJ2', 'DRD2', 'KATNA1', 'SDCBP', 'NIF3L1', 'HIST2H2BE', 'DUT', 'UBQLN2', 'TRIM43', 'TENC1', 'NEIL2', 'XCL2', 'NDUFA10', 'AP1S2', 'PLCL2', 'RAP2C', 'GALNT10', 'PTPRR', 'ZNF621', 'SYT9', 'STX6', 'TPBG', 'NPAT', 'PTK2', 'HEMK1', 'BTN2A1', 'GALM', 'ZNF226', 'PHACTR3', 'PTPN2', 'FIP1L1', 'LRRFIP1', 'KIAA1731', 'MEIS1', 'IGF1R', 'CRLF3', 'PPP3CB', 'RAB26', 'CSPG4', 'ZNF267', 'GPR155', 'PRPF18', 'CLDN9', 'GAMT', 'NOTCH2NL', 'COG5', 'NEK3', 'RGS12', 'DDHD1', 'PLEKHA1', 'ANK3', 'SEPP1', 'ZNF214', 'NFKB1', 'CCT4', 'PROCA1', 'P4HA2', 'FTCD', 'FGF20', 'ROR2', 'IFITM1', 'CREBL2', 'PITPNC1', 'PPP1CC', 'FZD7', 'NRN1', 'BFAR', 'TP73', 'IDH1', 'GCLC', 'SOX2', 'IL15', 'WDR45', 'RERG', 'CLN8', 'CLTC', 'ASB14', 'CLPX', 'ERN1', 'BAMBI', 'NETO1', 'KCNJ2', 'TM4SF1', 'PER1', 'FAM3C', 'CKS1B', 'GUCY1B3', 'APOBEC3C', 'AMICA1', 'BCL11B', 'PPP2R5E', 'DDX11']\n"
     ]
    }
   ],
   "source": [
    "# microarray as SC leg normalized, RUN This 1:1\n",
    "import pandas as pd\n",
    "SC_df = pd.read_csv(\"Skin-SunExposed(Lowerleg)_whitfield_micro_array_as_sc.csv\", sep='\\t', index_col = 0)\n",
    "\n",
    "new_list = list()\n",
    "for col,phase in zip(SC_df.columns, time_course_all_phases_list):\n",
    "    new_list.append(col+ \"_\"+phase)\n",
    "SC_df.columns = new_list\n",
    "print(SC_df.index)\n",
    "print(SC_df.columns)\n",
    "print(gene_names_with_micro_array_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['ABCF1', 'ALAS1', 'ANLN', 'ARL6IP1', 'ATF2', 'BCL6', 'BIRC5', 'BMP1',\n",
      "       'BTBD3', 'BUB1',\n",
      "       ...\n",
      "       'TUBB', 'UBE2C', 'UNG', 'USP1', 'VANGL1', 'VCAM1', 'VPS25', 'WDR76',\n",
      "       'ZC3HC1', 'ZNF367'],\n",
      "      dtype='object', length=252)\n",
      "(252, 930)\n"
     ]
    }
   ],
   "source": [
    "# 930 Dataset, RUN This 1:1\n",
    "\n",
    "import csv\n",
    "\n",
    "SC_df = pd.DataFrame()\n",
    "\n",
    "column_names = []\n",
    "with open('Dataset930.csv') as f:\n",
    "    reader = csv.reader(f, delimiter=\"\\t\")\n",
    "    next(reader)\n",
    "    counter = 0\n",
    "    full_row = []\n",
    "    gene_name = \"\"\n",
    "    for row in reader:\n",
    "        if (counter < 930):\n",
    "            phase = \"\"\n",
    "            if (row[1] == \"g0/g1\"):\n",
    "                phase = \"G1\"\n",
    "            elif (row[1] == \"s\"):\n",
    "                phase = \"S\"\n",
    "            elif (row[1] == \"g2/m\"):\n",
    "                phase = \"G2M\"\n",
    "            name = \"cell_\" + str(counter) + \"_\" + phase\n",
    "            column_names.append(name)\n",
    "        if (counter%930 == 0 and counter > 0):\n",
    "            SC_df[gene_name] = full_row\n",
    "            #print(full_row)\n",
    "            full_row = []\n",
    "            \n",
    "        full_row.append(2**float(row[9]))\n",
    "        gene_name = row[6]\n",
    "        \n",
    "        counter += 1\n",
    "SC_df = SC_df.T\n",
    "SC_df.columns = column_names\n",
    "print(SC_df.index)\n",
    "print(SC_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15 27 18\n",
      "10 genes used for cell A1_G1\n",
      "G1\n",
      "12 genes used for cell A2_G1\n",
      "G1\n",
      "16 genes used for cell A3_S\n",
      "S\n",
      "23 genes used for cell A4_S\n",
      "G1\n",
      "22 genes used for cell A5_G2M\n",
      "S\n",
      "18 genes used for cell A6_G2M\n",
      "G2M\n",
      "19 genes used for cell A7_G1\n",
      "G2M\n",
      "18 genes used for cell A8_G1\n",
      "G2M\n",
      "19 genes used for cell A9_S\n",
      "G1\n",
      "17 genes used for cell A10_S\n",
      "G1\n",
      "19 genes used for cell A11_G2M\n",
      "G2M\n",
      "19 genes used for cell A12_G2M\n",
      "G2M\n",
      "14 genes used for cell A2b_G1\n",
      "G2M\n",
      "18 genes used for cell B2_G1\n",
      "G2M\n",
      "19 genes used for cell A4b_S\n",
      "S\n",
      "23 genes used for cell B5_G2M\n",
      "S\n",
      "7 genes used for cell B6_G2M\n",
      "G2M\n",
      "12 genes used for cell A8b_G1\n",
      "G1\n",
      "9 genes used for cell B8_G1\n",
      "G1\n",
      "18 genes used for cell B9_S\n",
      "S\n",
      "18 genes used for cell B10_S\n",
      "G2M\n",
      "19 genes used for cell B11_G2M\n",
      "G1\n",
      "9 genes used for cell B12_G2M\n",
      "G1\n",
      "11 genes used for cell C1_G1\n",
      "G2M\n",
      "8 genes used for cell C2_G1\n",
      "S\n",
      "16 genes used for cell C3_S\n",
      "S\n",
      "20 genes used for cell A12b_G2M\n",
      "S\n",
      "10 genes used for cell C1b_G1\n",
      "G1\n",
      "21 genes used for cell C6_G2M\n",
      "S\n",
      "16 genes used for cell C7_G1\n",
      "G2M\n",
      "10 genes used for cell C8_G1\n",
      "G1\n",
      "20 genes used for cell C9_S\n",
      "S\n",
      "21 genes used for cell C10_S\n",
      "S\n",
      "19 genes used for cell C11_G2M\n",
      "G2M\n",
      "16 genes used for cell C12_G2M\n",
      "G1\n",
      "17 genes used for cell D1_G1\n",
      "S\n",
      "20 genes used for cell D2_G1\n",
      "S\n",
      "17 genes used for cell D3_S\n",
      "G2M\n",
      "20 genes used for cell C3b_S\n",
      "G2M\n",
      "20 genes used for cell C4b_S\n",
      "G2M\n",
      "18 genes used for cell D6_G2M\n",
      "S\n",
      "13 genes used for cell D7_G1\n",
      "S\n",
      "19 genes used for cell D8_G1\n",
      "S\n",
      "21 genes used for cell D9_S\n",
      "G1\n",
      "21 genes used for cell D10_S\n",
      "G1\n",
      "19 genes used for cell D11_G2M\n",
      "G2M\n",
      "22 genes used for cell D12_G2M\n",
      "G2M\n",
      "9 genes used for cell E1_G1\n",
      "S\n",
      "10 genes used for cell E2_G1\n",
      "G2M\n",
      "18 genes used for cell E3_S\n",
      "G2M\n",
      "19 genes used for cell E4_S\n",
      "S\n",
      "21 genes used for cell E5_G2M\n",
      "G2M\n",
      "18 genes used for cell E6_G2M\n",
      "S\n",
      "11 genes used for cell E7_G1\n",
      "S\n",
      "6 genes used for cell E8_G1\n",
      "G2M\n",
      "22 genes used for cell E9_S\n",
      "G2M\n",
      "15 genes used for cell E10_S\n",
      "S\n",
      "16 genes used for cell E11_G2M\n",
      "G1\n",
      "21 genes used for cell E12_G2M\n",
      "G2M\n",
      "19 genes used for cell C5b_G2M\n",
      "G2M\n",
      "17 genes used for cell F2_G1\n",
      "G2M\n",
      "18 genes used for cell F3_S\n",
      "G2M\n",
      "21 genes used for cell F4_S\n",
      "G2M\n",
      "17 genes used for cell F5_G2M\n",
      "G1\n",
      "20 genes used for cell F6_G2M\n",
      "G2M\n",
      "14 genes used for cell F7_G1\n",
      "S\n",
      "9 genes used for cell F8_G1\n",
      "G2M\n",
      "16 genes used for cell F9_S\n",
      "G2M\n",
      "20 genes used for cell F10_S\n",
      "S\n",
      "18 genes used for cell F11_G2M\n",
      "G2M\n",
      "20 genes used for cell F12_G2M\n",
      "G2M\n",
      "15 genes used for cell G1_G1\n",
      "G1\n",
      "18 genes used for cell G2_G1\n",
      "G2M\n",
      "18 genes used for cell G3_S\n",
      "G1\n",
      "20 genes used for cell G4_S\n",
      "G1\n",
      "17 genes used for cell G5_G2M\n",
      "G2M\n",
      "23 genes used for cell G6_G2M\n",
      "G1\n",
      "18 genes used for cell G7_G1\n",
      "S\n",
      "21 genes used for cell G9_S\n",
      "G2M\n",
      "21 genes used for cell G10_S\n",
      "S\n",
      "18 genes used for cell G11_G2M\n",
      "G2M\n",
      "17 genes used for cell G12_G2M\n",
      "S\n",
      "11 genes used for cell H1_G1\n",
      "S\n",
      "8 genes used for cell H2_G1\n",
      "G1\n",
      "20 genes used for cell H3_S\n",
      "S\n",
      "15 genes used for cell H4_S\n",
      "S\n",
      "19 genes used for cell H5_G2M\n",
      "G1\n",
      "18 genes used for cell H6_G2M\n",
      "G1\n",
      "18 genes used for cell H7_G1\n",
      "G2M\n",
      "11 genes used for cell H8_G1\n",
      "G2M\n"
     ]
    }
   ],
   "source": [
    "# this is version 5 v 5 v5 attempt at linear regression\n",
    "import statistics\n",
    "import datetime\n",
    "now = datetime.datetime.now()\n",
    "print(str(now.hour) + \" \" + str(now.minute) + \" \" + str(now.second))\n",
    "import sklearn\n",
    "from sklearn import linear_model\n",
    "SC_our_assignment_tupples_list = list()\n",
    "for cell_name in SC_df:\n",
    "    \n",
    "    # get a set of genes that are non-zero in that cell\n",
    "    count = 0\n",
    "    whitfield_subset_micro_array_df = pd.DataFrame(columns = range(0,91))\n",
    "    cell_cycle_genes_for_analysis = list()\n",
    "    cell_data = []\n",
    "    for gene_name, row in SC_df.iterrows():\n",
    "        if ((gene_name in gene_names_with_micro_array_data) and SC_df[cell_name][gene_name]):\n",
    "            count += 1\n",
    "            cell_cycle_genes_for_analysis.append(gene_name)\n",
    "            cell_data.append(SC_df[cell_name][gene_name])\n",
    "            whitfield_subset_micro_array_df.loc[gene_name] = gene_name_to_time_course_expressions_dict[gene_name][\"all\"]\n",
    "    print(str(count) + \" genes used for cell \" + cell_name)\n",
    "    #print(cell_cycle_genes_for_analysis)\n",
    "    cell_data_array = np.array(cell_data)\n",
    "    #print(cell_data_array[:5])\n",
    "    # huge, and pretty uniform MSE for all time points when fitting individual cells\n",
    "    mse_sum = 0\n",
    "    best_mse = math.inf\n",
    "    best_angle = -1\n",
    "    for i in range(0, 91): \n",
    "        lm = linear_model.LinearRegression()\n",
    "        #print(whitfield_subset_micro_array_df.iloc[:, [i]])\n",
    "        lm.fit(whitfield_subset_micro_array_df.iloc[:, [i]], cell_data_array)\n",
    "        #print(cell_data_array)\n",
    "        #print(lm.predict(whitfield_subset_micro_array_df.iloc[:, [i]]))\n",
    "        mse = np.mean((cell_data_array - lm.predict(whitfield_subset_micro_array_df.iloc[:, [i]])) **2)\n",
    "        #print(\"MSE of using just a column fit for the first column:\")\n",
    "        #print(mse)\n",
    "        mse_sum += mse\n",
    "        if (mse < best_mse):\n",
    "            best_mse = mse\n",
    "            best_angle = time_course_all_angles_list[i]\n",
    "    \n",
    "    # using the entire micro-array for a single fit results in a higher average MSE, but test it out later\n",
    "    # to see if by chance the accuracy is higher... - did better for the bulk, at least!!!!\n",
    "    #print(\"MSE of using a jsut a column for each column, Average\")\n",
    "    #print(mse_sum/91)\n",
    "    lm = linear_model.LinearRegression()\n",
    "    lm.fit(whitfield_subset_micro_array_df, cell_data_array)\n",
    "    #print(lm.intercept_)\n",
    "    #print(lm.coef_)\n",
    "    #print(whitfield_subset_micro_array_df.iloc[:, [0]])\n",
    "    mse_sum = 0\n",
    "    best_mse = math.inf\n",
    "    best_angle = -1\n",
    "    for i in range(0, 91): \n",
    "    #    print(i)\n",
    "    #    print(whitfield_subset_micro_array_df.iloc[:, [i]][:5])\n",
    "        new = whitfield_subset_micro_array_df.iloc[:, [i]] * lm.coef_[i] + lm.intercept_\n",
    "    #    #print(new)\n",
    "        mse = np.mean((cell_data_array - np.array(new)) **2)\n",
    "        mse_sum += mse\n",
    "        if (mse < best_mse):\n",
    "            best_mse = mse\n",
    "            best_angle = time_course_all_angles_list[i]\n",
    "    #print(\"MSE of using a whole array fit, Average\")\n",
    "    #print(mse_sum/91)\n",
    "    \n",
    "    angle = best_angle\n",
    "    predicted_phase = -1\n",
    "    # Angles - Version 2\n",
    "    # angles version 2\n",
    "    if angle >= 0 and angle <= 0.8:\n",
    "        predicted_phase = \"G1\" #G1_angles += 1\n",
    "    elif angle > 0.8 and angle <= 2.5:\n",
    "        predicted_phase = \"S\" #S_angles += 1\n",
    "    elif angle > 2.5 and angle <= 5.4:\n",
    "        predicted_phase = \"G2M\" #G2M_angles += 1\n",
    "    elif angle > 5.4:\n",
    "        predicted_phase = \"G1\" #G1_angles += 1\n",
    "    # Angles - Version 1\n",
    "    # angles version 2\n",
    "    #if angle >= 0 and angle <= 2:\n",
    "    #    predicted_phase = \"S\"\n",
    "    #elif angle > 2 and angle <= 4.8:\n",
    "    #    predicted_phase = \"G2M\"\n",
    "    #elif angle > 4.8 and angle <= 5.9:\n",
    "    #    predicted_phase = \"G1\"\n",
    "    #elif angle > 5.9:\n",
    "    #    predicted_phase = \"S\"\n",
    "    \n",
    "    print(predicted_phase)\n",
    "    SC_our_assignment_tupples_list.append((cell_name, predicted_phase))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "F1_score\n",
      "0.3880184067088499\n",
      "num S: 29\n",
      "num G2: 30\n",
      "num G1: 31\n"
     ]
    }
   ],
   "source": [
    "S_tp = 0\n",
    "S_fp = 0\n",
    "S_fn = 0\n",
    "G2M_tp = 0\n",
    "G2M_fp = 0\n",
    "G2M_fn = 0\n",
    "G1_tp = 0\n",
    "G1_fp = 0\n",
    "G1_fn = 0\n",
    "num_s = 0\n",
    "num_g2 = 0\n",
    "num_g1 = 0\n",
    " \n",
    "#SC_our_assignment_tupples_list = [(\"1_G1\", \"G2M\"), (\"2_S\", \"G1\"), (\"3_G2M\",\"G2M\")]\n",
    "for (cell_name, predicted_phase) in SC_our_assignment_tupples_list:\n",
    "    if (\"_S\" in cell_name):\n",
    "        num_s += 1\n",
    "        if (predicted_phase == \"S\"):\n",
    "            S_tp += 1\n",
    "        else:\n",
    "            S_fn += 1\n",
    "            if (predicted_phase == \"G2M\"):\n",
    "                G2M_fp += 1\n",
    "            elif (predicted_phase == \"G1\"):\n",
    "                G1_fp += 1\n",
    "    elif (\"_G2\" in cell_name):   \n",
    "        num_g2 += 1\n",
    "        if (predicted_phase == \"G2M\"):\n",
    "            G2M_tp += 1        \n",
    "        else:\n",
    "            G2M_fn += 1\n",
    "            if (predicted_phase == \"S\"):\n",
    "                S_fp += 1\n",
    "            elif (predicted_phase == \"G1\"):\n",
    "                G1_fp += 1\n",
    "    elif (\"_G1\" in cell_name):\n",
    "        num_g1 += 1\n",
    "        if (predicted_phase == \"G1\"):\n",
    "            G1_tp += 1\n",
    "        else:\n",
    "            G1_fn += 1\n",
    "            if (predicted_phase == \"S\"):\n",
    "                S_fp += 1\n",
    "            elif (predicted_phase == \"G2M\"):\n",
    "                G2M_fp += 1 \n",
    "\n",
    "S_precision = S_tp / (S_tp + S_fp) if S_tp else 0\n",
    "S_recall = S_tp / (S_tp + S_fn) if S_tp else 0\n",
    "G2M_precision = G2M_tp / (G2M_tp + G2M_fp) if G2M_tp else 0\n",
    "G2M_recall = G2M_tp / (G2M_tp + G2M_fn) if G2M_tp else 0\n",
    "G1_precision = G1_tp / (G1_tp + G1_fp) if G1_tp else 0\n",
    "G1_recall = G1_tp / (G1_tp + G1_fn) if G1_tp else 0\n",
    "precision_avg = (S_precision + G2M_precision + G1_precision)/3\n",
    "recall_avg = (S_recall + G2M_recall + G1_recall)/3\n",
    "f1_score = 2*((precision_avg*recall_avg)/(precision_avg+recall_avg))\n",
    "print(\"F1_score\")\n",
    "print(f1_score)\n",
    "print(\"num S: \" + str(num_s))\n",
    "print(\"num G2: \" + str(num_g2))\n",
    "print(\"num G1: \" + str(num_g1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene1    100\n",
      "gene2     38\n",
      "Name: 0, dtype: int64\n",
      "[50, 2]\n"
     ]
    }
   ],
   "source": [
    "test_df = pd.DataFrame()\n",
    "test_df[\"gene1\"] = [100,200]\n",
    "test_df[\"gene2\"] = [38, 12]\n",
    "print(test_df.loc[0])\n",
    "test_cell = [50, 2]\n",
    "import sklearn\n",
    "from sklearn import linear_model\n",
    "lm = linear_model.LinearRegression()\n",
    "print(test_cell)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-27.419354838709666\n",
      "[0.77419355]\n",
      "2.524354896707238e-29\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/moamin/anaconda3/lib/python3.6/site-packages/ipykernel/__main__.py:1: FutureWarning: reshape is deprecated and will raise in a subsequent release. Please use .values.reshape(...) instead\n",
      "  if __name__ == '__main__':\n",
      "/home/moamin/anaconda3/lib/python3.6/site-packages/ipykernel/__main__.py:4: FutureWarning: reshape is deprecated and will raise in a subsequent release. Please use .values.reshape(...) instead\n"
     ]
    }
   ],
   "source": [
    "lm.fit(test_df.loc[0].reshape(-1, 1), test_cell)\n",
    "print(lm.intercept_)\n",
    "print(lm.coef_)\n",
    "mse = np.mean((test_cell - lm.predict(test_df.loc[0].reshape(-1, 1))) **2)\n",
    "print(mse)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
